lib/srv/discovery/fetchers/aws-sync/reconcile.go

```diff
@@ -24,640 +24,271 @@ import (
+		reconcile(old.UserAttachedPolicies, new.UserAttachedPolicies, userAttchPolKey, userAttchPolWrap),
+		reconcile(old.UserGroups, new.UserGroups, userGroupKey, userGroupWrap),
+		reconcile(old.Groups, new.Groups, groupKey, groupWrap),
+		reconcile(old.GroupInlinePolicies, new.GroupInlinePolicies, grpInlinePolKey, grpInlinePolWrap),
+		reconcile(old.GroupAttachedPolicies, new.GroupAttachedPolicies, grpAttchPolKey, grpAttchPolWrap),
+		reconcile(old.Policies, new.Policies, policyKey, policyWrap),
+		reconcile(old.Instances, new.Instances, instanceKey, instanceWrap),
+		reconcile(old.S3Buckets, new.S3Buckets, s3bucketKey, s3bucketWrap),
+		reconcile(old.Roles, new.Roles, roleKey, roleWrap),
+		reconcile(old.RoleInlinePolicies, new.RoleInlinePolicies, roleInlinePolKey, roleInlinePolWrap),
+		reconcile(old.RoleAttachedPolicies, new.RoleAttachedPolicies, roleAttchPolKey, roleAttchPolWrap),
+		reconcile(old.InstanceProfiles, new.InstanceProfiles, instanceProfKey, instanceProfWrap),
+		reconcile(old.EKSClusters, new.EKSClusters, eksClusterKey, eksClusterWrap),
+		reconcile(old.AssociatedAccessPolicies, new.AssociatedAccessPolicies, assocAccPolKey, assocAccPolWrap),
+		reconcile(old.AccessEntries, new.AccessEntries, accessEntryKey, accessEntryWrap),
+		reconcile(old.RDSDatabases, new.RDSDatabases, rdsDbKey, rdsDbWrap),
+		reconcile(old.SAMLProviders, new.SAMLProviders, samlProvKey, samlProvWrap),
+		reconcile(old.OIDCProviders, new.OIDCProviders, oidcProvKey, oidcProvWrap),
+	}
+	for _, res := range reconciledResources {
+		upsert.Resources = append(upsert.Resources, res.upsert.Resources...)
+		delete.Resources = append(delete.Resources, res.delete.Resources...)
 	}
-
 	return upsert, delete
 }
 
-type reconcileIntermeditateResult struct {
-	upsert, delete *accessgraphv1alpha.AWSResourceList
+type reconcilePair struct {
+	upsert, delete *tag.AWSResourceList
 }
 
-func instancesUniqueKeyFunc(instance *accessgraphv1alpha.AWSInstanceV1) string {
-	return fmt.Sprintf("%s;%s", instance.InstanceId, instance.Region)
+func deduplicateSlice[T any](s []T, key func(T) string) []T {
+	out := make([]T, 0, len(s))
+	seen := make(map[string]struct{})
+	for _, v := range s {
+		if _, ok := seen[key(v)]; ok {
+			continue
+		}
+		seen[key(v)] = struct{}{}
+		out = append(out, v)
+	}
+	return out
 }
 
-func reconcileInstances(old []*accessgraphv1alpha.AWSInstanceV1, new []*accessgraphv1alpha.AWSInstanceV1) *reconcileIntermeditateResult {
-	upsert, delete := &accessgraphv1alpha.AWSResourceList{}, &accessgraphv1alpha.AWSResourceList{}
+func reconcile[T protoreflect.ProtoMessage](
+	oldItems []T,
+	newItems []T,
+	keyFn func(T) string,
+	wrapFn func(T) *tag.AWSResource,
+) *reconcilePair {
+	// Remove duplicates from the new items
+	newItems = deduplicateSlice(newItems, keyFn)
+	upsertRes := newResourceList()
+	deleteRes := newResourceList()
+
+	// Return upsert if there are no old items, and vice versa
+	if len(newItems) == 0 || len(oldItems) == 0 {
```

Comment:
I'm not sure if this optimizes anything. If there are no elements then nothing below will execute anyway. IMO having only one path make function simpler to understand.

Reply:
It is a little confusing for not much less lines of code. Will revert it back.